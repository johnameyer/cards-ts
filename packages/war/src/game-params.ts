import { Serializable } from '@cards-ts/core';

export interface GameParams {
    readonly [key: string]: Serializable;
    
    // TODO readonly limitWarToMinCards: boolean;
    readonly maxBattles: number;
}
