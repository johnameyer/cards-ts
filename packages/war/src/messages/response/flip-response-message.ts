import { Message } from '@cards-ts/core';

export class FlipResponseMessage extends Message {
    readonly type = 'flip-response';

    constructor(public readonly data?: any) {
        super([ 'Flipped a card' ]);
    }
}
