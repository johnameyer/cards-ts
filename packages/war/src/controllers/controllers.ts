import { HiddenHandsControllerProvider, DefaultControllerKeys, DefaultControllers, SystemHandlerParams, DeckControllerProvider, ControllersProviders, UnwrapProviders, ValidatedProviders } from "@cards-ts/core";
import { GameStates } from "../game-states";
import { GameParams } from "../game-params";
import { WarControllerProvider } from "./war-controller";
import { GameHandlerParams } from "../game-handler-params";
import { ResponseMessage } from "../messages/response-message";

type TypedDefaultControllers = DefaultControllers<GameParams, typeof GameStates, ResponseMessage, GameHandlerParams & SystemHandlerParams>;

export const buildProviders = () => {
    const providers = {
        deck: new DeckControllerProvider(1, true, false),
        hand: new HiddenHandsControllerProvider(),
        war: new WarControllerProvider()
    };
    return providers as Omit<ValidatedProviders<typeof providers & ControllersProviders<TypedDefaultControllers>>, DefaultControllerKeys>;
};

export type GameControllers = UnwrapProviders<ReturnType<typeof buildProviders>>;

export type Controllers = GameControllers & TypedDefaultControllers;