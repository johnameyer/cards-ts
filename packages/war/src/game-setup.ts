import { GenericGameSetup, Intermediary } from '@cards-ts/core';
import { GameParams } from './game-params';

export class GameSetup implements GenericGameSetup<GameParams> {
    getDefaultParams(): GameParams {
        return {
            // TODO limitWarToMinCards: true
            maxBattles: 200,
        };
    }
    
    async setupForIntermediary(host: Intermediary): Promise<GameParams> {
        const [ _, resultsPromise ] = host.form(
            { type: 'input', message: [ 'How many battles before declaring a stalemate? (default 200)' ] },
        );

        const results = await resultsPromise;

        const maxBattles = Number(results[0]) || 200;

        return {
            maxBattles,
        };
    }

    
    verifyParams(params: GameParams): { readonly maxBattles?: string; } {
        const errors: { maxBattles?: string | undefined; } = {};
        try {
            if(!Number(params.maxBattles) || Number(params.maxBattles) <= 0) {
                throw new Error();
            }
        } catch (e) {
            errors.maxBattles = 'Max battles must be an positive number.';
        }
        return errors;
    }

    getYargs(): {[key: string]: import('yargs').Options} {
        return {
            maxBattles: { alias: 'n', description: 'How many battles before declaring a stalemate', type: 'number', default: 200 },
        };
    }

    setupForYargs(params: any): GameParams {
        return {
            maxBattles: Number(params.maxBattles),
        };
    }
}
