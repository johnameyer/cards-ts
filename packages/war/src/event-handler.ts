import { EventHandlerInterface, GenericGameState } from '@cards-ts/core';
import { Controllers } from './controllers/controllers';
import { ResponseMessage } from './messages/response-message';

export class EventHandler implements EventHandlerInterface<Controllers, ResponseMessage> {
    merge(controllers: Controllers, sourceHandler: number, incomingEvent: ResponseMessage): void {
        switch(incomingEvent.type) {
            case 'flip-response': {
                controllers.data.setDataFor(sourceHandler, incomingEvent.data);
                controllers.waiting.removePosition(sourceHandler);
                return;
            }
            case 'data-response': {
                controllers.data.setDataFor(sourceHandler, incomingEvent.data);
                return;
            }
        }
    }

    validateEvent(_controllers: Controllers, _sourceHandler: number, event: ResponseMessage): ResponseMessage | undefined {
        switch(event.type) {
            case 'flip-response': {
                return event;
            }
            case 'data-response': {
                return event;
            }
        }
    }
}