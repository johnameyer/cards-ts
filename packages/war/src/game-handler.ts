import { ControllerHandlerState, Handler, HandlerResponsesQueue } from '@cards-ts/core';
import { Controllers } from './controllers/controllers';
import { GameHandlerParams } from './game-handler-params';
import { ResponseMessage } from './messages/response-message';

export type HandlerData = ControllerHandlerState<Controllers>;

export abstract class GameHandler implements Handler<GameHandlerParams, HandlerData, ResponseMessage> {
    abstract handleFlip(handlerData: HandlerData, responsesQueue: HandlerResponsesQueue<ResponseMessage>): void | Promise<void>;
}
