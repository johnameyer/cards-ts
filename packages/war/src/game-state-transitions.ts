import { SpacingMessage, GenericGameState, GenericGameStateTransitions } from '@cards-ts/core';
import { GameStates } from './game-states';
import { FlippedMessage, GameOverMessage, StalemateMessage, WonBattleMessage, WonWarMessage } from './messages/status';
import { Controllers } from './controllers/controllers';

type GameState = GenericGameState<Controllers>;

export class GameStateTransitions implements GenericGameStateTransitions<typeof GameStates, Controllers> {
    public get() {
        return {
            [GameStates.START_GAME]: this.startGame,
            [GameStates.START_BATTLE]: this.startBattle,
            [GameStates.START_FLIP]: this.startFlip,
            [GameStates.WAIT_FOR_FLIP]: this.waitForFlip,
            [GameStates.HANDLE_FLIP]: this.handleFlip,
            [GameStates.END_BATTLE]: this.endBattle,
            [GameStates.START_WAR]: this.startWar,
            [GameStates.END_WAR]: this.endWar,
            [GameStates.END_GAME]: this.endGame,
        };
    }

    startGame(controllers: Controllers) {
        controllers.deck.resetDeck();
        controllers.hand.dealOut(false);

        controllers.state.set(GameStates.START_BATTLE);
    }

    startBattle(controllers: Controllers) {
        controllers.war.resetPlayed();

        controllers.state.set(GameStates.START_FLIP);
        
        if(controllers.war.battleCountAndOne > controllers.params.get().maxBattles) {
            controllers.state.set(GameStates.END_GAME);
        }
    }

    startFlip(controllers: Controllers) {
        controllers.state.set(GameStates.WAIT_FOR_FLIP);
    }

    waitForFlip(controllers: Controllers) {
        controllers.players.handlerCallAll('flip');
        
        controllers.state.set(GameStates.HANDLE_FLIP);
    }

    handleFlip(controllers: Controllers) {
        controllers.war.flipCards();

        if(controllers.war.getMaxPlayed() === 1) {
            controllers.state.set(GameStates.END_BATTLE);
        } else if((controllers.war.getMaxPlayed() - 1) % 4 === 0 || controllers.hand.numberOfPlayersWithCards() !== controllers.names.get().length) {
            controllers.state.set(GameStates.END_WAR);
        } else {
            controllers.state.set(GameStates.START_FLIP);
        }
    }

    endBattle(controllers: Controllers) {
        const flipped = controllers.war.getLastFlipped();
        for(let i = 0; i < flipped.length; i++) {
            controllers.players.messageAll(new FlippedMessage(controllers.names.get(i), flipped[i]));
            // TODO move these kinds of messages into controllers
        }
        
        const max = controllers.war.battleWinner();
        
        if(max === -1) {
            controllers.state.set(GameStates.START_WAR);
            return;
        }
        controllers.war.givePlayedTo(max);

        controllers.players.messageAll(new WonBattleMessage(controllers.names.get()[max]));
        controllers.players.messageAll(new SpacingMessage());

        if(controllers.hand.numberOfPlayersWithCards() === 1) {
            controllers.state.set(GameStates.END_GAME);
        } else {
            controllers.state.set(GameStates.START_BATTLE);
        }
    }

    startWar(controllers: Controllers) {
        controllers.state.set(GameStates.START_FLIP);
    }

    endWar(controllers: Controllers) {
        const flipped = controllers.war.getLastFlipped();
        for(let i = 0; i < flipped.length; i++) {
            controllers.players.messageAll(new FlippedMessage(controllers.names.get(i), flipped[i]));
        }

        const max = controllers.war.battleWinner();
        
        if(max === -1) {
            controllers.state.set(GameStates.START_WAR);
            return;
        }
        controllers.war.givePlayedTo(max);
        
        controllers.players.messageAll(new WonWarMessage(controllers.names.get(max)));
        controllers.players.messageAll(new SpacingMessage());

        if(controllers.hand.numberOfPlayersWithCards() === 1) {
            controllers.state.set(GameStates.END_GAME);
        } else {
            controllers.state.set(GameStates.START_BATTLE);
        }
    }

    endGame(controllers: Controllers) {
        if(controllers.hand.numberOfPlayersWithCards() > 1) {
            controllers.players.messageAll(new StalemateMessage());
        } else {
            controllers.players.messageAll(new GameOverMessage(controllers.names.get(controllers.hand.handWithMostCards())));
        }
        controllers.completed.complete();
    }
}
