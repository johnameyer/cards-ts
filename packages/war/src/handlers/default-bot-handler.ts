import { HandlerResponsesQueue, Message } from "@cards-ts/core";
import { GameHandler } from "../game-handler";
import { HandlerData } from "../handler-data";
import { FlipResponseMessage } from "../messages/response";
import { ResponseMessage } from "../messages/response-message";

export class DefaultBotHandler extends GameHandler {
    handleFlip(handlerData: HandlerData, responsesQueue: HandlerResponsesQueue<ResponseMessage>): void | Promise<void> {
        responsesQueue.push(new FlipResponseMessage());
    }
}