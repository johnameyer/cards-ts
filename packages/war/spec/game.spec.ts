import { expect } from 'chai';
import { IntermediaryHandler } from '../src/index.js';
import { eventHandler } from '../src/event-handler.js';
import { GameSetup } from '../src/game-setup.js';
import { gameStateTransitions } from '../src/game-state-transitions.js';
import { DefaultBotHandler } from '../src/handlers/default-bot-handler.js';
import { GameHandler } from '../src/game-handler.js';
import { GameParams } from '../src/game-params.js';
import { buildProviders } from '../src/controllers/controllers.js';
import { ArrayMessageHandler, buildGameFactory, Card, DeckControllerProvider, HandlerChain } from '@cards-ts/core';
import { FlippedMessage, StalemateMessage, WonBattleMessage } from '../src/messages/status/index.js';
import { FlipResponseMessage } from '../src/messages/response/flip-response-message.js';
import { StatusMessage } from '../../can-i-have-that/src/messages/status-message.js';

describe('game', () => {
    // TODO can we build this more simply i.e. deterministic deck controller?
    const mockProviders = () => ({
        ...buildProviders(),
        deck: new DeckControllerProvider(1, false, false), // note shouldShuffle false
    });

    const factory = buildGameFactory(
        gameStateTransitions,
        eventHandler,
        // TODO can we avoid specifying the setup and intermediary since they are not critical to the driver construction?
        new GameSetup(),
        intermediary => new IntermediaryHandler(intermediary),
        () => new DefaultBotHandler(),
        mockProviders,
    );

    it('works as expected', async () => {
        const messageHandlers = Array.from({length: 2}, () => new ArrayMessageHandler<StatusMessage>());

        const gameHandler: () => GameHandler =  () => ({
            handleFlip: (_state, responses) => {
                responses.push(new FlipResponseMessage());
            },
        });

        // @ts-expect-error
        const players = Array.from(messageHandlers, messageHandler => new HandlerChain([ messageHandler, gameHandler() ]));

        const params: GameParams = {
            ...new GameSetup().getDefaultParams(),
            maxBattles: 2,
        };

        const names = Array.from(messageHandlers, (_, i) => String.fromCharCode(65 + i));

        const driver = factory.getGameDriver(players, params, names);

        driver.resume();

        for(const messageHandler of messageHandlers) {
            expect(messageHandler.arr).to.be.empty;
        }

        expect(driver.getState().completed).to.be.false;

        driver.handleSyncResponses();
        driver.resume();

        const firstFlips = [
            new FlippedMessage('A', Card.fromString('2D', 0)), // TODO can we avoid specifying the deck number?
            new FlippedMessage('B', Card.fromString('3D', 0)),
            new WonBattleMessage('B'),
        ];


        for(const messageHandler of messageHandlers) {
            expect(messageHandler.arr).to.deep.include.ordered.members(firstFlips);
            messageHandler.clear();
        }

        expect(driver.getState().completed).to.be.false;

        driver.handleSyncResponses();
        driver.resume();

        const secondFlips = [
            new FlippedMessage('A', Card.fromString('4D', 0)),
            new FlippedMessage('B', Card.fromString('5D', 0)),
            new WonBattleMessage('B'),
            new StalemateMessage(),
        ];

        for(const messageHandler of messageHandlers) {
            expect(messageHandler.arr).to.deep.include.ordered.members(secondFlips);
            messageHandler.clear();
        }

        expect(driver.getState().completed).to.be.true;

        // TODO test multiflipping - would require more test setup
    });
});
