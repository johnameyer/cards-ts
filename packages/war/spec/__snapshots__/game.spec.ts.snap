// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game has a matching snapshot 1`] = `
Array [
  "A flipped 2 of Diamonds",
  "B flipped 3 of Diamonds",
  "B won the battle",
  "A flipped 4 of Diamonds",
  "B flipped 5 of Diamonds",
  "B won the battle",
  "Game went over the max number of battles",
]
`;

exports[`game has a matching snapshot 2`] = `
Array [
  "A flipped 2 of Diamonds",
  "B flipped 3 of Diamonds",
  "B won the battle",
  "A flipped 4 of Diamonds",
  "B flipped 5 of Diamonds",
  "B won the battle",
  "Game went over the max number of battles",
]
`;
